source/extensions/transport_sockets/alts/alts_tsi_handshaker.cc

```diff
@@ -0,0 +1,169 @@
+#include "source/extensions/transport_sockets/alts/alts_tsi_handshaker.h"
+
+#include <algorithm>
+#include <cstdint>
+#include <cstring>
+#include <memory>
+#include <string>
+#include <utility>
+#include <vector>
+
+#include "absl/memory/memory.h"
+#include "absl/status/status.h"
+#include "absl/strings/str_format.h"
+#include "src/core/lib/iomgr/exec_ctx.h"
+#include "src/core/tsi/alts/zero_copy_frame_protector/alts_zero_copy_grpc_protector.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace TransportSockets {
+namespace Alts {
+
+using ::grpc::gcp::HandshakerResp;
+
+constexpr std::size_t AltsAes128GcmRekeyKeyLength = 44;
+
+std::unique_ptr<AltsTsiHandshaker>
+AltsTsiHandshaker::createForClient(std::shared_ptr<grpc::Channel> handshaker_service_channel) {
+  return absl::WrapUnique(new AltsTsiHandshaker(/*is_client=*/true, handshaker_service_channel));
+}
+
+std::unique_ptr<AltsTsiHandshaker>
+AltsTsiHandshaker::createForServer(std::shared_ptr<grpc::Channel> handshaker_service_channel) {
+  return absl::WrapUnique(new AltsTsiHandshaker(/*is_client=*/false, handshaker_service_channel));
+}
+
+AltsTsiHandshaker::AltsTsiHandshaker(bool is_client,
+                                     std::shared_ptr<grpc::Channel> handshaker_service_channel)
+    : is_client_(is_client), handshaker_service_channel_(handshaker_service_channel) {}
+
+absl::Status AltsTsiHandshaker::next(void* handshaker, const unsigned char* received_bytes,
+                                     size_t received_bytes_size, OnNextDone on_next_done) {
+  // Argument and state checks.
+  if (handshaker == nullptr || (received_bytes == nullptr && received_bytes_size > 0) ||
+      on_next_done == nullptr) {
+    return absl::InvalidArgumentError("Invalid nullptr argument to AltsTsiHandshaker::Next.");
+  }
+  if (is_handshake_complete_) {
+    return absl::InternalError("Handshake is already complete.");
+  }
+
+  // Get a handshake message from the handshaker service.
+  absl::string_view in_bytes(reinterpret_cast<const char*>(received_bytes), received_bytes_size);
```

Comment:
do you mean to drop the unsigned here?

Reply:
Fixed, replaced with `absl::Span<const uint8_t>`.